### ISSUE TITLE
Incorrect usage of ReplicaRunning()

### ISSUE THREAD

**Issue by ggunson** [openark/orchestrator#763]
In investigating https://github.com/github/orchestrator/pull/761 with the concern that `RESET MASTER` might be being issued before replication had fully stopped, I found that the check for whether replication was running or not wasn't complete.

The check used in some parts of orchestrator to tell that replication was running on a server was:

https://github.com/github/orchestrator/blob/a58ad0fdd55b42fddf4553f7db02002db9d6c9d6/go/inst/instance.go#L256-L259

The function returns true if the host is a replica and both sql and IO threads are "Yes". It can then return false if one thread is still running (or the IO thread is "Connecting") as well as when both threads are stopped. So ReplicaRunning() is an invalid function to use to check that replication has 100% stopped.

Currently the function is being used in a few places, but the places where I see a possible concern are:

- [SyncReplicaRelayLogs()](https://github.com/github/orchestrator/blob/master/go/agent/instance_topology_agent.go#L27) -- if the IO thread is still running then the relay log contents are still being written to, and if the SQL thread is still on then it may still be applying relay log contents.

- [StartSlaveUntilMasterCoordinates()](https://github.com/github/orchestrator/blob/a58ad0fdd55b42fddf4553f7db02002db9d6c9d6/go/inst/instance_topology_dao.go#L479) -- orc might try running `start slave until` on a host where one of the threads is already running, which in the console results in only a warning and **no change** to the `until` values. 

- [EnableMasterSSL()](https://github.com/github/orchestrator/blob/master/go/inst/instance_topology_dao.go#L578) -- `CHANGE MASTER` could try executing on a replica with one of the threads running (which should error in this case)*

- [ChangeMasterTo()](https://github.com/github/orchestrator/blob/master/go/inst/instance_topology_dao.go#L605) -- similar to above

- [ResetSlave()](https://github.com/github/orchestrator/blob/master/go/inst/instance_topology_dao.go#L706) -- [RESET SLAVE](https://dev.mysql.com/doc/refman/5.7/en/reset-slave.html) could be run on a replica where one of the threads is still running, which I believe would result in an error but you'll want to double-check

- [ResetMaster()](https://github.com/github/orchestrator/blob/master/go/inst/instance_topology_dao.go#L739) -- RESET MASTER could happen while, say, the SQL thread is still running, so writes could happen after the RESET MASTER that you didn't intend (and if you're going to change gtid_purged or something after this it'll be incorrect)

- [DetachReplica()](https://github.com/github/orchestrator/blob/master/go/inst/instance_topology_dao.go#L867) -- orc could detach a replica from a topology with the SQL thread still running and get the logged coordinates for the `CHANGE MASTER` wrong

- [ReattachReplica()](https://github.com/github/orchestrator/blob/master/go/inst/instance_topology_dao.go#L901) -- less likely as the previous function but it still just checks for both threads running

*Note with MySQL 5.7 `CHANGE MASTER` will run in some cases when replication is still partially up (either IO or SQL threads running). See https://dev.mysql.com/doc/refman/5.7/en/change-master-to.html starting at "In MySQL 5.7.4 and later". 

/cc https://github.com/github/orchestrator/pull/762 
/cc https://github.com/github/orchestrator/pull/761
/cc @github/database-infrastructure

**Comment by ggunson** [comment]
Also re: `CHANGE MASTER` command:

> In MySQL 5.7.4 and later, relay logs are preserved if at least one of the slave SQL thread and the slave I/O thread is running; if both threads are stopped, all relay log files are deleted unless at least one of RELAY_LOG_FILE or RELAY_LOG_POS is specified.

So in 5.7 and newer `CHANGE MASTER` won't delete relay logs if one of the replication threads is still running. 

Edited to add: This could mean potentially replicating things you don't want (if you were trying to move the replica out from a host writing errant transactions), or result in replication lag upon reconnection to a new master since the SQL thread would need to go through the old relay logs before starting on the new ones the IO thread captures from the new master. 

https://dev.mysql.com/doc/refman/5.7/en/change-master-to.html

**Comment by shlomi-noach** [comment]
Thank you -- I'll review all the above functions.

https://github.com/github/orchestrator/pull/762, merged a few days ago, introduces `NoReplicationThreadRunning()`:

https://github.com/github/orchestrator/pull/762/files#diff-6479b8ed383c25fd0ebea6f81a75cffdR262

Example usage:

https://github.com/github/orchestrator/pull/762/files#diff-bfc0f008baba7b60fb6a291f728701c3R556

#762 also introduced a fix to checking thread states, (previously a `Connecting` state could implicitly be interpreted as a `No`):

https://github.com/github/orchestrator/pull/762/files#diff-3844df65307d41626850bed65098717cR251

**Comment by ggunson** [comment]
Also found in orchestrator-client's `is_replicating()`:

https://github.com/github/orchestrator/blob/a58ad0fdd55b42fddf4553f7db02002db9d6c9d6/resources/bin/orchestrator-client#L384-L390

**Comment by shlomi-noach** [comment]
Thanks. `is_replicating` is behaving as expected, and should return `true` when both replication threads are running. I'll perhaps introduce a `is_not_replicating` command to complement it.

**Comment by ggunson** [comment]
>  I'll perhaps introduce a is_not_replicating command to complement it.

That was my hope, I should have been clear that I mentioned the function in orchestrator-client because it does the same check as ReplicaRunning() and has the same concerns.

**Comment by shlomi-noach** [comment]
Cool. Unfortunately I also realize the implementation for `ReplicaNotRunning()` is also incorrect, because it yet doesn't take into consideration a `Connecting` thread state. I'll introduce a better logic in a new PR.

**Comment by shlomi-noach** [comment]
MySQL documentation reference: https://dev.mysql.com/doc/refman/8.0/en/show-slave-status.html

**Comment by shlomi-noach** [comment]
Tracking fixes:

- [x]    SyncReplicaRelayLogs() -- if the IO thread is still running then the relay log contents are still being written to, and if the SQL thread is still on then it may still be applying relay log contents.
- [x]    StartSlaveUntilMasterCoordinates() -- orc might try running start slave until on a host where one of the threads is already running, which in the console results in only a warning and no change to the until values.
- [x]    EnableMasterSSL() -- CHANGE MASTER could try executing on a replica with one of the threads running (which should error in this case)*
- [x]    ChangeMasterTo() -- similar to above
- [x]    ResetSlave() -- RESET SLAVE could be run on a replica where one of the threads is still running, which I believe would result in an error but you'll want to double-check
- [x]    ResetMaster() -- RESET MASTER could happen while, say, the SQL thread is still running, so writes could happen after the RESET MASTER that you didn't intend (and if you're going to change gtid_purged or something after this it'll be incorrect)
- [x]    DetachReplica() -- orc could detach a replica from a topology with the SQL thread still running and get the logged coordinates for the CHANGE MASTER wrong
- [x]    ReattachReplica() -- less likely as the previous function but it still just checks for both threads running

**Comment by shlomi-noach** [comment]
For `ChangeMasterTo()`, `ResetSlave()`, `ResetMaster()`, the operation should also be permitted if the instance is not a replica in the first place.
For `EnableMasterSSL()` I'm unsure but including it as well.
